### PR TITLE
implements rlp node decoder

### PR DESCRIPTION
**Pull request by kjezek**
This PR implements a decoder from an RLP encoded byte slice into a Node. 

It uses the intermediate rlp.Item format to convert from the byte slice and then it converts this format to a node.

### CODE REVIEW

File: go/database/mpt/decoder_test.go [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+
+func Test_isCompactPathLeafNode(t *testing.T) {
+	tests := []struct {
+		path   []byte
+		isLeaf bool
+	}{
+		{[]byte{0b_0000_0000}, false},
+		{[]byte{0b_0001_0000}, false},
+		{[]byte{0b_0010_0000}, true},
+		{[]byte{0b_0011_0000}, true},
```
Comment (HerbertJordan) [inline go/database/mpt/decoder_test.go:277]: nit: if you want to stress-test this, test all possible 256 values for the first byte.

**Review by HerbertJordan — APPROVED** [review]
Looks good, thanks!

File: go/database/mpt/decoder.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+		nextNode[n] = 0xF // terminal marker
+	}
+
+	return &ExtensionNode{path: CreatePathFromNibbles(path), nextHash: nextNode, nextIsEmbedded: embedded}, nil
+}
+
+// decodeLeafNodeFromRlp decodes a leaf node from RLP-encoded data.
+// A leaf node can be either a value node or an account node.
+// The node type is distinguished by the length of the payload.
+// The value not has a payload of size <= common.ValueSize,
```
[the remark this answers is not in the document]
Comment (kjezek) [reply go/database/mpt/decoder.go:98]: I think `node` was intended :D

File: go/database/mpt/decoder.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+// In the compat encoding, the first nibble of the path contains the oddness of the path,
+// and if the node is lead or not.
+// The encoding is as follows:
+// - 0b_0000_0000 (0x00): extension node, even path
+// - 0b_0001_xxxx (0x1_): extension node, odd path
+// - 0b_0010_0000 (0x20): leaf node, even path
+// - 0b_0011_xxxx (0x3_): leaf node, odd path
+// for more see:
+// https://arxiv.org/pdf/2108.05513/1000 sec 4.1
+func isCompactPathLeafNode(path []byte) bool {
```
[the remark this answers is not in the document]
Comment (kjezek) [reply go/database/mpt/decoder.go]: ok, renamed

File: go/database/mpt/decoder_test.go [hunk 4]
```diff
@@ -0,0 +1,374 @@
...
+}
+
+func TestDecoder_DecodeEmbeddedNode_CanDecode(t *testing.T) {
+	value := []byte{0x11, 0x22, 0x33, 0x44}
+	valueNode := rlp.List{Items: []rlp.Item{rlp.String{Str: []byte{0x20, 0x12, 0x34}}, rlp.String{Str: value}}}
+	valueNodeRlp := rlp.Encode(valueNode)
+
+	var embeddedAsHash common.Hash
+	copy(embeddedAsHash[:], valueNodeRlp)
+	embeddedAsHash[len(valueNodeRlp)] = 0xF // add termination mark
```
[the remark this answers is not in the document]
Comment (kjezek) [reply go/database/mpt/decoder_test.go]: yes,  updated

File: go/database/mpt/decoder.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (kjezek) [reply go/database/mpt/decoder.go]: fixed

File: go/database/mpt/decoder_test.go (hunk 1 above)
Comment (kjezek) [reply go/database/mpt/decoder_test.go]: good idea, updated

**Review by petr-hanzl — APPROVED** [review]
lgtm. thanks

**Review by HerbertJordan — APPROVED** [review]

**Pull request merged**